### PR TITLE
added-order-entity

### PR DESCRIPTION
**Pull request by margosalii**

### CODE REVIEW

**Comment by margosalii** [comment]
image
image
I need help with this task. When receiving all user orders, the BookId is always null. I checked in the database, but all the data is displayed correctly there, but in Postman and Swagger it is null. 
Also, please advise how to properly clean the Shopping Cart. After all, nothing changes when I use shoppingCart.getCartItems().clear();

File: src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        return orderItemRepository.findAllByOrderId(id)
+            .stream()
+            .map(orderItemMapper::toDto)
+            .collect(Collectors.toSet());
+    }
+
+    @Transactional
+    private void saveOrderItem(OrderItem orderItem) {
+        orderItemRepository.save(orderItem);
+    }
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java]: Why do you need this? transaction in transaction)))

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java (hunk 1 above)
Comment (margosalii) [reply src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java]: > Why do you need this? transaction in transaction)))

So, if the method that called saveOrderItem already has a Transactional annotation over it, then i no longer need to use it over saveOrderItem? Is this what you meant? I just thought that this annotation should be used in all methods that create or change information in the DB

File: src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    @Override
+    public OrderItemDto getItemByOrderIdAndItemId(Long orderId, Long id) {
+        OrderItem orderItem = orderItemRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find item by id: " + id));
+
+        if (orderItem.getOrder().getId().equals(orderId)) {
+            return orderItemMapper.toDto(orderItem);
+        }
```
Comment (sarakhmen) [inline src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java]: Please reduce the java-code logic that could be present on the BD side. You can query order items by id and order id, and it'll be much quicker

File: src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        throw new EntityNotFoundException("Can't find item by order id: " + orderId);
+    }
+
+    @Override
+    public Set<OrderItemDto> getItemsByOrderId(Long id) {
+        return orderItemRepository.findAllByOrderId(id)
+            .stream()
+            .map(orderItemMapper::toDto)
+            .collect(Collectors.toSet());
```
Comment (sarakhmen) [inline src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java]: mupstruct can be used to map List of items to List of other items

Comment (sarakhmen) [reply src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java]: fix the same issue in all places

File: src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            orderItem.setBook(cartItem.getBook());
+            orderItem.setPrice(cartItem.getBook().getPrice());
+            orderItem.setQuantity(cartItem.getQuantity());
+            orderItem.setOrder(order);
+            orderItems.add(orderItem);
+        }
+        return orderItems;
+    }
+
+    private BigDecimal countTotal(Set<OrderItem> orderItems) {
```
Comment (sarakhmen) [inline src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java]:
```suggestion
    private BigDecimal calculateTotal(Set<OrderItem> orderItems) {
```

**Review by sarakhmen — CHANGES_REQUESTED** [review]
let's slightly improve your solution :)

File: src/main/java/mate/academy/store/model/Order.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE orders SET is_deleted = true WHERE id=?")
+@SQLRestriction("is_deleted=false")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/store/model/Order.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/store/model/Order.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    private Status status;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/store/model/Order.java]:
```suggestion
    private Status status = Status.PENDING;
```

File: src/main/java/mate/academy/store/model/Order.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Column(nullable = false)
+    private BigDecimal total;
+
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/store/model/Order.java]:
```suggestion
    private LocalDateTime orderDate = LocalDateTime.now();
```

File: src/main/java/mate/academy/store/model/Order.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Column(nullable = false)
+    private String shippingAddress;
+
+    @OneToMany(mappedBy = "order")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/store/model/Order.java]:
```suggestion
    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/mate/academy/store/model/OrderItem.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE order_items SET is_deleted = true WHERE id=?")
+@SQLRestriction("is_deleted=false")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/store/model/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/store/model/OrderItem.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/store/model/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/store/repository/order/OrderRepository.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package mate.academy.store.repository.order;
+
+import java.util.Set;
+import mate.academy.store.model.Order;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface OrderRepository extends JpaRepository<Order, Long> {
+    Set<Order> findAllByUserId(Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/store/repository/order/OrderRepository.java:10]: `@EntityGraph(attributePaths = "orderItems")`

File: src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+        Order order = new Order();
+        Set<OrderItem> orderItems = createSetOfOrderItems(order, shoppingCart);
+
+        order.setOrderItems(orderItems);
+        order.setTotal(calculateTotal(orderItems));
+        order.setShippingAddress(requestOrderDto.getShippingAddress());
+        order.setUser(shoppingCart.getUser());
+        order.setOrderDate(LocalDateTime.now());
+        order.setStatus(Status.PENDING);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java]:
```suggestion
```

see comment in entity

File: src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    @Override
+    public List<ResponseOrderDto> getAllUsersOrders(Long userId) {
+        return orderRepository.findAllByUserId(userId)
+            .stream()
+            .map(orderMapper::toDto)
+            .toList();
+    }
+
+    @Override
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/store/service/order/impl/OrderServiceImpl.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**